File: Controllers/QuestionnaireController.cs
using MVC_C_sharp.Models;
using MVC_C_sharp.Data;
using MVC_C_sharp.Repositories;

namespace MVC_C_sharp.Controllers
{
    public class QuestionnaireController
    {
        private readonly QuestionnaireRepository _questionnaireRepository;
        private readonly QuestionRepository _questionRepository;
        private readonly ReponseRepository _reponseRepository;
        private readonly UtilisateurRepository _utilisateurRepository;
        private readonly AdminLogRepository _adminLogRepository;

        public QuestionnaireController()
        {
            DatabaseInitializer.EnsureAdminAndLogTables();

            _questionnaireRepository = new QuestionnaireRepository();
            _questionRepository = new QuestionRepository();
            _reponseRepository = new ReponseRepository();
            _utilisateurRepository = new UtilisateurRepository();
            _adminLogRepository = new AdminLogRepository();
        }

        public List<Questionnaire> GetAllQuestionnaires()
        {
            return _questionnaireRepository.GetAll();
        }

        public List<Questionnaire> GetQuestionnaires(int utilisateurId)
        {
            return _questionnaireRepository.GetAllByUtilisateur(utilisateurId);
        }

        public Questionnaire? GetQuestionnaire(int id)
        {
            return _questionnaireRepository.GetById(id);
        }

        public int CreateQuestionnaire(string nom, string theme, int utilisateurId)
        {
            int questionnaireId = _questionnaireRepository.Create(nom, theme, utilisateurId);

            if (questionnaireId > 0)
            {
                string pseudo = GetUtilisateurPseudo(utilisateurId);
                _adminLogRepository.CreateLog(
                    utilisateurId,
                    "QUESTIONNAIRE_CREE",
                    $"Le questionnaire '{nom}' (theme : {theme}) a ete cree par '{pseudo}'."
                );
            }

            return questionnaireId;
        }

        public bool UpdateQuestionnaire(int id, string nom, string theme)
        {
            return _questionnaireRepository.Update(id, nom, theme);
        }

        public bool DeleteQuestionnaire(int id, int utilisateurId)
        {
            var questionnaire = _questionnaireRepository.GetById(id);
            bool deleted = _questionnaireRepository.Delete(id);

            if (deleted)
            {
                string pseudo = GetUtilisateurPseudo(utilisateurId);
                string nom = questionnaire?.Nom ?? $"Questionnaire #{id}";
                string theme = questionnaire?.Theme ?? "Inconnu";

                _adminLogRepository.CreateLog(
                    utilisateurId,
                    "QUESTIONNAIRE_SUPPRIME",
                    $"Le questionnaire '{nom}' (theme : {theme}) a ete supprime par '{pseudo}'."
                );
            }

            return deleted;
        }

        public void LogQuestionnaireCompletion(int utilisateurId, Questionnaire questionnaire, int score, int total)
        {
            string pseudo = GetUtilisateurPseudo(utilisateurId);
            double pourcentage = total > 0 ? (double)score / total * 100 : 0;

            _adminLogRepository.CreateLog(
                utilisateurId,
                "QUESTIONNAIRE_TERMINE",
                $"Le questionnaire '{questionnaire.Nom}' (theme : {questionnaire.Theme}) a ete termine par '{pseudo}' avec un score de {score}/{total} ({pourcentage:F0}%)."
            );
        }

        public List<Question> GetQuestions(int questionnaireId)
        {
            return _questionRepository.GetAllByQuestionnaire(questionnaireId);
        }

        public Question? GetQuestion(int id)
        {
            return _questionRepository.GetById(id);
        }

        public int CreateQuestion(int questionnaireId, string libelle, string typeReponse, bool? reponseVraiFaux)
        {
            int numero = _questionRepository.GetNextNumero(questionnaireId);
            return _questionRepository.Create(questionnaireId, numero, libelle, typeReponse, reponseVraiFaux);
        }

        public bool UpdateQuestion(int id, string libelle, string typeReponse, bool? reponseVraiFaux)
        {
            return _questionRepository.Update(id, libelle, typeReponse, reponseVraiFaux);
        }

        public bool DeleteQuestion(int id)
        {
            return _questionRepository.Delete(id);
        }

        public List<Reponse> GetReponses(int questionId)
        {
            return _reponseRepository.GetAllByQuestion(questionId);
        }

        public int CreateReponse(int questionId, string valeur, bool estCorrecte)
        {
            return _reponseRepository.Create(questionId, valeur, estCorrecte);
        }

        public bool UpdateReponse(int id, string valeur, bool estCorrecte)
        {
            return _reponseRepository.Update(id, valeur, estCorrecte);
        }

        public bool DeleteReponse(int id)
        {
            return _reponseRepository.Delete(id);
        }

        public bool DeleteAllReponses(int questionId)
        {
            return _reponseRepository.DeleteAllByQuestion(questionId);
        }

        private string GetUtilisateurPseudo(int utilisateurId)
        {
            var utilisateur = _utilisateurRepository.GetById(utilisateurId);
            return utilisateur?.Pseudo ?? $"User #{utilisateurId}";
        }
    }
}

File: web/controllers/QuestionnaireController.php
<?php
require_once __DIR__ . '/../models/Questionnaire.php';
require_once __DIR__ . '/../models/Question.php';
require_once __DIR__ . '/../models/Reponse.php';

class QuestionnaireController {
    
    public static function getAll() {
        return Questionnaire::getAll();
    }

    public static function getMine($utilisateurId) {
        return Questionnaire::getAllByUtilisateur($utilisateurId);
    }

    public static function getPlayHistory($utilisateurId, $limit = 20) {
        return Questionnaire::getPlayHistoryByUtilisateur($utilisateurId, $limit);
    }

    public static function getById($id) {
        return Questionnaire::getById($id);
    }

    public static function create($nom, $theme, $utilisateurId) {
        if (empty($nom)) {
            return ['success' => false, 'message' => 'Le nom est obligatoire'];
        }
        if (empty($theme)) {
            return ['success' => false, 'message' => 'Le thème est obligatoire'];
        }

        $id = Questionnaire::create($nom, $theme, $utilisateurId);
        if ($id) {
            return ['success' => true, 'message' => 'Questionnaire créé', 'id' => $id];
        }
        return ['success' => false, 'message' => 'Erreur lors de la création'];
    }

    public static function update($id, $nom, $theme, $utilisateurId) {
        $questionnaire = Questionnaire::getById($id);
        if (!$questionnaire || $questionnaire->utilisateurId != $utilisateurId) {
            return ['success' => false, 'message' => 'Questionnaire non trouvé ou non autorisé'];
        }

        if (empty($nom)) {
            return ['success' => false, 'message' => 'Le nom est obligatoire'];
        }

        if (Questionnaire::update($id, $nom, $theme)) {
            return ['success' => true, 'message' => 'Questionnaire mis à jour'];
        }
        return ['success' => false, 'message' => 'Erreur lors de la mise à jour'];
    }

    public static function delete($id, $utilisateurId) {
        $questionnaire = Questionnaire::getById($id);
        if (!$questionnaire || $questionnaire->utilisateurId != $utilisateurId) {
            return ['success' => false, 'message' => 'Questionnaire non trouvé ou non autorisé'];
        }

        if (Questionnaire::delete($id)) {
            return ['success' => true, 'message' => 'Questionnaire supprimé'];
        }
        return ['success' => false, 'message' => 'Erreur lors de la suppression'];
    }

    public static function getQuestions($questionnaireId) {
        return Question::getAllByQuestionnaire($questionnaireId);
    }

    public static function getQuestion($id) {
        return Question::getById($id);
    }

    public static function addQuestion($questionnaireId, $libelle, $typeReponse, $reponseVraiFaux = null, $reponses = []) {
        if (empty($libelle)) {
            return ['success' => false, 'message' => 'Le libellé est obligatoire'];
        }

        $numero = Question::getNextNumero($questionnaireId);
        $questionId = Question::create($questionnaireId, $numero, $libelle, $typeReponse, $reponseVraiFaux);

        if ($questionId && $typeReponse === 'ListeValeurs') {
            foreach ($reponses as $reponse) {
                Reponse::create($questionId, $reponse['valeur'], $reponse['estCorrecte']);
            }
        }

        if ($questionId) {
            return ['success' => true, 'message' => 'Question ajoutée', 'id' => $questionId];
        }
        return ['success' => false, 'message' => 'Erreur lors de l\'ajout'];
    }

    public static function updateQuestion($id, $libelle, $typeReponse, $reponseVraiFaux = null, $reponses = []) {
        if (empty($libelle)) {
            return ['success' => false, 'message' => 'Le libellé est obligatoire'];
        }

        if (Question::update($id, $libelle, $typeReponse, $reponseVraiFaux)) {
            Reponse::deleteAllByQuestion($id);
            
            if ($typeReponse === 'ListeValeurs') {
                foreach ($reponses as $reponse) {
                    Reponse::create($id, $reponse['valeur'], $reponse['estCorrecte']);
                }
            }
            
            return ['success' => true, 'message' => 'Question mise à jour'];
        }
        return ['success' => false, 'message' => 'Erreur lors de la mise à jour'];
    }

    public static function deleteQuestion($id) {
        if (Question::delete($id)) {
            return ['success' => true, 'message' => 'Question supprimée'];
        }
        return ['success' => false, 'message' => 'Erreur lors de la suppression'];
    }

    public static function getReponses($questionId) {
        return Reponse::getAllByQuestion($questionId);
    }

    public static function checkAnswer($questionId, $answer) {
        $question = Question::getById($questionId);
        if (!$question) {
            return ['correct' => false];
        }

        if ($question->typeReponse === 'VraiFaux') {
            $correct = ($answer === 'true' && $question->reponseVraiFaux) || 
                       ($answer === 'false' && !$question->reponseVraiFaux);
            return ['correct' => $correct];
        } else {
            $reponses = Reponse::getAllByQuestion($questionId);
            foreach ($reponses as $reponse) {
                if ($reponse->id == $answer && $reponse->estCorrecte) {
                    return ['correct' => true];
                }
            }
            return ['correct' => false];
        }
    }

    public static function checkAndStoreAnswer($utilisateurId, $questionnaireId, $questionId, $answer) {
        $question = Question::getById($questionId);
        if (!$question) {
            return ['correct' => false];
        }

        $result = self::checkAnswer($questionId, $answer);
        $isCorrect = $result['correct'] === true;

        $reponseBool = null;
        if ($question->typeReponse === 'VraiFaux') {
            $reponseBool = $answer === 'true' ? 1 : 0;
        }

        Questionnaire::saveUserAnswer(
            $utilisateurId,
            $questionnaireId,
            $questionId,
            (string)$answer,
            $reponseBool,
            $isCorrect
        );

        return ['correct' => $isCorrect];
    }

    public static function getPedagogicalStats() {
        return [
            'questionnairesByTheme' => Questionnaire::getQuestionnaireCountByTheme(),
            'weeklyConnections' => Questionnaire::getWeeklyUserConnections(),
            'successByTheme' => Questionnaire::getSuccessRateByTheme()
        ];
    }

    public static function trackQuestionnaireAccess($utilisateurId, $questionnaireId) {
        return Questionnaire::trackQuestionnaireAccess($utilisateurId, $questionnaireId);
    }

    public static function publish($id, $utilisateurId) {
        $questionnaire = Questionnaire::getById($id);
        if (!$questionnaire || $questionnaire->utilisateurId != $utilisateurId) {
            return ['success' => false, 'message' => 'Questionnaire non trouvé ou non autorisé'];
        }

        if ($questionnaire->estPublie) {
            return ['success' => false, 'message' => 'Ce questionnaire est déjà publié'];
        }

        if ($questionnaire->nombreQuestions < 1) {
            return ['success' => false, 'message' => 'Ajoutez au moins une question avant de publier'];
        }

        if (Questionnaire::publish($id, $utilisateurId)) {
            return ['success' => true, 'message' => 'Questionnaire publié avec succès'];
        }

        return ['success' => false, 'message' => 'Erreur lors de la publication'];
    }
}
